panic: runtime error: invalid memory address or nil pointer dereference

Hello,
I have an ubuntu VM with influxdb version 1.3.5 installed on it:
```
root@hostname:~# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description: Ubuntu 14.04 LTS
Release: 14.04
Codename: trusty
```
InfluxDB shell version: 1.3.5
I configured a retention policy 4 days via influx shell:
```
> show retention policies
name duration shardGroupDuration replicaN default
---- -------- ------------------ -------- -------
autogen 0s 168h0m0s 1 false
telegraf 168h0m0s 24h0m0s 1 true
```

Since then the influxdb service fails every few hours about memory issue:
```
[I] 2017-09-04T03:50:19Z retention policy shard deletion check commencing service=retention
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xb8 pc=0x820c20]
goroutine 314 [running]:
github.com/influxdata/influxdb/tsdb.(*Shard).UnloadIndex(0xc420466000)
/root/go/src/github.com/influxdata/influxdb/tsdb/shard.go:401 +0x30
github.com/influxdata/influxdb/tsdb.(*Store).DeleteShard(0xc4201f6f20, 0x43, 0x43, 0xc4209dacd8)
/root/go/src/github.com/influxdata/influxdb/tsdb/store.go:441 +0x55
github.com/influxdata/influxdb/services/retention.(*Service).deleteShards(0xc420146be0)
/root/go/src/github.com/influxdata/influxdb/services/retention/service.go:123 +0x75c
created by github.com/influxdata/influxdb/services/retention.(*Service).Open
/root/go/src/github.com/influxdata/influxdb/services/retention/service.go:46 +0x181
```

So I have to restart the service every few hours as workaround.
Can you assist to fix it? Does the retention policy configuration looks bad?
The VM had 8GB of memory, I increased it to 16GB but the issue still the same.
Thank you!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -48,6 +48,7 @@
 - [#8766](https://github.com/influxdata/influxdb/pull/8766): Fix deadlock when calling `SeriesIDsAllOrByExpr`
 - [#8638](https://github.com/influxdata/influxdb/issues/8638): Fix `influx_inspect export` so it skips missing files.
 - [#8770](https://github.com/influxdata/influxdb/pull/8770): Reduce how long it takes to walk the varrefs in an expression.
+- [#8787](https://github.com/influxdata/influxdb/issues/8787): panic: runtime error: invalid memory address or nil pointer dereference.
 
 ## v1.3.4 [unreleased]
 

--- a/tsdb/shard.go
+++ b/tsdb/shard.go
@@ -112,7 +112,6 @@ type Shard struct {
 	path    string
 	walPath string
 	id      uint64
-	wg      sync.WaitGroup
 
 	database        string
 	retentionPolicy string
@@ -350,11 +349,10 @@ func (s *Shard) close(clean bool) error {
 	default:
 		close(s.closing)
 	}
-	s.wg.Wait()
 
 	if clean {
 		// Don't leak our shard ID and series keys in the index
-		s.UnloadIndex()
+		s.unloadIndex()
 	}
 
 	err := s.engine.Close()
@@ -403,6 +401,13 @@ func (s *Shard) LastModified() time.Time {
 
 // UnloadIndex removes all references to this shard from the DatabaseIndex
 func (s *Shard) UnloadIndex() {
+	if err := s.ready(); err != nil {
+		return
+	}
+	s.unloadIndex()
+}
+
+func (s *Shard) unloadIndex() {
 	s.index.RemoveShard(s.id)
 }
 

